ext/drstatecmp/drstatecmp.c

```diff
@@ -0,0 +1,691 @@
+     * last instruction if the bb falls through.
+     */
+    instr_t *term_inst_copy_bb = instrlist_last_app(copy_bb);
+    bool preinsert = true;
+    if (!instr_is_cti(term_inst_copy_bb))
+        preinsert = false;
+    labels->term = drstatecmp_insert_label(drcontext, copy_bb, term_inst_copy_bb,
+                                           DRSTATECMP_LABEL_TERM, preinsert);
+
+    /* Delete the terminating instruction of the original bb (if any) to let the original
+     * bb fall through to its copy for re-execution.
+     */
+    instr_t *term_inst = instrlist_last_app(bb);
+    if (instr_is_cti(term_inst)) {
+        instrlist_remove(bb, term_inst);
+        instr_destroy(drcontext, term_inst);
+    }
+
+    /* Append the instructions of the bb copy to the original bb. */
+    instrlist_append(bb, labels->copy_bb_start);
+    /* Empty and destroy the bb copy (but not its instructions) since it is not needed
+     * anymore.
+     */
+    instrlist_init(copy_bb);
+    instrlist_destroy(drcontext, copy_bb);
+}
+
+static void
+drstatecmp_save_state_call(int for_cmp)
+{
+    void *drcontext = dr_get_current_drcontext();
+    drstatecmp_saved_states_t *pt =
+        (drstatecmp_saved_states_t *)drmgr_get_tls_field(drcontext, tls_idx);
+
+    dr_mcontext_t *mcontext = NULL;
+    if (for_cmp)
+        mcontext = &pt->saved_state_for_cmp;
+    else
+        mcontext = &pt->saved_state_for_restore;
+    mcontext->size = sizeof(*mcontext);
+    mcontext->flags = DR_MC_ALL;
+    dr_get_mcontext(drcontext, mcontext);
+}
+
+static void
+drstatecmp_save_state(void *drcontext, instrlist_t *bb, instr_t *instr, bool for_cmp)
+{
+    dr_insert_clean_call(drcontext, bb, instr, (void *)drstatecmp_save_state_call,
+                         false /* fpstate */, 1, OPND_CREATE_INT32((int)for_cmp));
+}
+
+static void
+drstatecmp_restore_state_call(void)
+{
+    void *drcontext = dr_get_current_drcontext();
+    drstatecmp_saved_states_t *pt =
+        (drstatecmp_saved_states_t *)drmgr_get_tls_field(drcontext, tls_idx);
+
+    dr_mcontext_t *mcontext = &pt->saved_state_for_restore;
+    mcontext->size = sizeof(*mcontext);
+    mcontext->flags = DR_MC_ALL;
+    dr_set_mcontext(drcontext, mcontext);
+}
+
+static void
+drstatecmp_restore_state(void *drcontext, instrlist_t *bb, instr_t *instr)
+{
+    dr_insert_clean_call(drcontext, bb, instr, (void *)drstatecmp_restore_state_call,
+                         false /* fpstate */, 0);
+}
+
+static void
+drstatecmp_report_error(const char *msg, void *tag)
+{
+    if (ops.error_callback != NULL) {
+        (*ops.error_callback)(msg, tag);
+    } else {
+        byte *bb_start_pc = dr_fragment_app_pc(tag);
+        void *drcontext = dr_get_current_drcontext();
+        instrlist_t *bb = decode_as_bb(drcontext, bb_start_pc);
+        dr_fprintf(STDERR, "Application basic block where mismatch detected: \n");
+        instrlist_disassemble(drcontext, bb_start_pc, bb, STDERR);
```

Comment:
Having the mcontext values would also be useful: how about a TODO or XXX comment for now.

Reply:
Done.